### ISSUE TITLE
Convert v_hashmap to std::unordered_map

### ISSUE THREAD

**Issue by peterychang** [VowpalWabbit/vowpal_wabbit#2019]
Performance testing shows that std::unordered_map (with max_load_factor = 0.25) is at least as performant, and in many cases much more performant than v_hashmap. It also uses a standardized interface and is an overall safer data structure.

This task is to track the conversion of v_hashmap to std::unordered_map in VW

Performance test results follow:
I ran 5 different map implementations (dense_hash_map and unordered_maps configured as closely to v_hashmap’s behavior as possible)
1.	v_hashmap with calloc’d substring keys
2.	v_hashmap with std::string keys
3.	v_hashmap with calloc’d substrings, no calls to uniform_hash (vector index for the source strings are used instead)
4.	google dense_hash_map with std::string keys, 25% max load factor and no size reduction on clear
5.	std::unordered_map with std::string keys and 25% max load factor

I also ran those implementations through 3 scenarios
1.	insert + clear 312 strings; large enough to force a resize in all implementations
2.	insert + clear 10 strings; probably small enough to not resize any implementation
3.	get inserted values

TL;DR:
•	For writes, dense_hash_map gives a 25-100x performance boost over v_hashmap. Unordered_map gives a 5-10x performance boost.
•	For reads, dense_hash_map doubles the throughput over v_hashmap. Unordered_map has comparable (< 10% better) performance.
•	Uniform_hash adds very little overhead
```
2019-08-12 09:55:59
Running ./perf_tests
Run on (40 X 2195 MHz CPU s)
Load Average: 0.52, 0.58, 0.59
----------------------------------------------------------------------------------
Benchmark                                        Time             CPU   Iterations
----------------------------------------------------------------------------------
insert_and_clear_v_hashmap                   23996 ns        24065 ns        29867
insert_and_clear_v_hashmap_string            28075 ns        27623 ns        24889
insert_and_clear_v_hashmap_nohash            22088 ns        21973 ns        32000
insert_and_clear_g_densemap                    854 ns          837 ns       746667
insert_and_clear_std_hashmap                  5440 ns         5469 ns       100000
insert_and_clear_v_hashmap_small              2153 ns         2176 ns       344615
insert_and_clear_v_hashmap_small_string       2688 ns         2668 ns       263529
insert_and_clear_v_hashmap_small_nohash       2001 ns         1995 ns       344615
insert_and_clear_g_densemap_small             25.0 ns         25.1 ns     28000000
insert_and_clear_std_hashmap_small             197 ns          195 ns      3446154
v_hashmap_get_values                          9097 ns         8998 ns        74667
v_hashmap_get_values_string                   9115 ns         8998 ns        74667
g_densemap_get_values                         5078 ns         5000 ns       100000
std_hashmap_get_values                        8290 ns         8371 ns        89600
```

**Comment by peterychang** [comment]
Currently, the 2 key types we use in v_hashmap are uint64_t and substring. Unordered_map should use std::string instead of substring.

However, looking at the code v_hashmap is inextricably tied to substrings. Moreover, these substrings contain pointers to dynamically allocated memory, which require custom cleanup code.

There are 2 approaches we can take here:
1. Make a small change to use unordered_map. This would require conversion between std::string and substring as well as custom allocation/deallocation code. It would, however be a relatively small change
2. Refactor the code to use boost::string_view instead of substring. These provide an easily convertible interface between strings and string references (string_view). Code would need to be refactored to ensure all strings referenced by string_view outlive the string_views themselves.

The first option will result in a fairly large performance hit, since we would need to deallocate memory and construct a new string many times.
The second option will result in a single large and complex change

**Comment by peterychang** [comment]
Performance testing around string views have shown comparable performance to substrings in most cases, better in the memory allocation case

```
2019-08-20 10:41:31
Running ./perf_tests
Run on (40 X 2195 MHz CPU s)
Load Average: 0.52, 0.58, 0.59
--------------------------------------------------------------------------------
Benchmark                                      Time             CPU   Iterations
--------------------------------------------------------------------------------
substring_from_string                        194 ns          193 ns      3733333 
string_view_from_string                      254 ns          257 ns      2800000
substring_from_string_prefix_suffix          497 ns          500 ns      1000000 
string_view_from_string_prefix_suffix        378 ns          377 ns      1866667
substring_from_string_copy                  8934 ns         8998 ns        74667 
string_from_string                          3262 ns         3296 ns       213333
substring_to_substring                       256 ns          257 ns      2800000 
string_view_to_string_view                   193 ns          193 ns      3733333
substring_to_string                         3356 ns         3369 ns       213333 
string_view_to_string                       3769 ns         3767 ns       186667
substring_compare                           3451 ns         3453 ns       203636
string_view_compare                          346 ns          345 ns      1947826
```

Per John, string_views are an acceptable alternative to substrings, and the migration should be completed in a single PR

**Comment by jackgerrits** [comment]
v_hashmap is gone already
